image/preseed/preseed.go

```diff
@@ -39,9 +54,153 @@ type preseedOpts struct {
+	trusted                  = sysdb.Trusted()
+)
+
+func MockTrusted(mockTrusted []asserts.Assertion) (restore func()) {
+	prevTrusted := trusted
+	trusted = mockTrusted
+	return func() {
+		trusted = prevTrusted
+	}
+}
+
+func writePreseedAssertion(opts *preseedOpts, artifactDigest []byte) error {
+	keypairMgr, err := getKeypairManager()
+	if err != nil {
+		return err
+	}
+
+	key := opts.PreseedSignKey
+	if key == "" {
+		key = `default`
+	}
+	privKey, err := keypairMgr.GetByName(key)
+	if err != nil {
+		// TRANSLATORS: %q is the key name, %v the error message
+		return fmt.Errorf(i18n.G("cannot use %q key: %v"), key, err)
+	}
+
+	sysDir := filepath.Join(opts.PrepareImageDir, "system-seed")
+	sd, err := seedOpen(sysDir, opts.SystemLabel)
+	if err != nil {
+		return err
+	}
+
+	bs := asserts.NewMemoryBackstore()
+	adb, err := asserts.OpenDatabase(&asserts.DatabaseConfig{
+		Trusted:        trusted,
+		KeypairManager: keypairMgr,
+		Backstore:      bs,
+	})
+	if err != nil {
+		return err
+	}
+
+	commitTo := func(b *asserts.Batch) error {
+		return b.CommitTo(adb, nil)
+	}
+
+	if err := sd.LoadAssertions(adb, commitTo); err != nil {
+		return err
+	}
+	model := sd.Model()
+
+	tm := timings.New(nil)
+	if err := sd.LoadMeta("run", tm); err != nil {
+		return err
+	}
+
+	snaps := []interface{}{}
+	addSnap := func(sn *seed.Snap) {
+		preseedSnap := map[string]interface{}{}
+		preseedSnap["name"] = sn.SnapName()
+		if sn.ID() != "" {
+			preseedSnap["id"] = sn.ID()
+			preseedSnap["revision"] = sn.PlaceInfo().SnapRevision().String()
+		}
+		snaps = append(snaps, preseedSnap)
+	}
+
+	modeSnaps, err := sd.ModeSnaps("run")
+	if err != nil {
+		return err
+	}
+	essSnaps := sd.EssentialSnaps()
+	if err != nil {
+		return err
+	}
+	for _, ess := range essSnaps {
+		addSnap(ess)
+	}
+	for _, msnap := range modeSnaps {
+		addSnap(msnap)
+	}
+
+	base64Digest, err := asserts.EncodeDigest(crypto.SHA3_384, artifactDigest)
+	if err != nil {
+		return err
+	}
+	headers := map[string]interface{}{
+		"type":              "preseed",
+		"authority-id":      model.AuthorityID(),
+		"series":            "16",
+		"brand-id":          model.BrandID(),
+		"model":             model.Model(),
+		"system-label":      opts.SystemLabel,
+		"artifact-sha3-384": base64Digest,
+		"timestamp":         time.Now().UTC().Format(time.RFC3339),
+		"revision":          "1",
+		"snaps":             snaps,
+	}
+
+	signedAssert, err := adb.Sign(asserts.PreseedType, headers, nil, privKey.PublicKey().ID())
+	if err != nil {
+		return fmt.Errorf("cannot sign preseed asertion: %v", err)
+	}
+
+	tsto, err := newToolingStoreFromModel(model, "")
+	if err != nil {
+		return err
+	}
+	newFetcher := func(save func(asserts.Assertion) error) asserts.Fetcher {
+		return tsto.AssertionFetcher(adb, save)
+	}
+
+	f := seedwriter.MakeRefAssertsFetcher(newFetcher)
+	if err := f.Save(signedAssert); err != nil {
+		return fmt.Errorf("cannot save assertion: %v", err)
```

Comment:
s/save/fetch/ as if there will be error they will be about the assertions we don't have